enhancements/monitoring/alert-overrides.md

```diff
@@ -0,0 +1,281 @@
+rules to adjust thresholds, or to simply add and modify labels to aide in
+routing alert notifications, as well as a supported method of adding additional
+alerting rules.
+
+Users currently have the ability to add additional rules from a technical
+standpoint -- by simply creating new `PrometheusRule` objects -- but this is
+[explicitly unsupported][1] at the moment.  The reason being that these
+resources are handled directly by the prometheus-operator, making it difficult
+to perform OpenShift specific defaulting and validation.  A concrete example
+being that there is no way to prevent users from creating new recording rules
+that may overwhelm the platform Prometheus instance.
+
+The goal of this proposal is to enable the addition of custom alerting rules,
+and limited ability to override labeling of existing platform alerting rules via
+resources under the direct control of the cluster-monitoring-operator.
+
+### Goals
+
+- Give users a supported and documented method to add additional rules, in a way
+  that allows the cluster-monitoring-operator to provide its own defaulting,
+  validation, telemetry, etc.
+- Allow users to apply relabeling operations to alerting rules shipped by the
+  platform.
+- Resources shipped by operators remain unmodified.  No action needed from
+  OpenShift developers.
+
+### Non-Goals
+
+- Only alerting rules are available for modification.  Not recording rules or
+  other monitoring configuration.
+- This phase of the enhancement does not address modifying alerting rule
+  expressions or thresholds.  This may be addressed by a later enhancement,
+  possibly with something similar to the [time series as alert thresholds][2]
+  pattern.
+  - Users still have the option of creating a new `AlertingRule` resource based
+    on an existing platform rule, making their modifications there, and
+    silencing the original alert.
+
+## Proposal
+
+The cluster-monitoring-operator will be extended with two new resources types:
+`AlertingRule` and `AlertRelabelConfig`, both in the `monitoring.openshift.io`
+group.
+
+The new `AlertingRule` type will allow additional user-defined alerting rules,
+but without the ability to create recording rules provided by the existing
+`PrometheusRule` resource type. Example:
+
+```yaml
+apiVersion: v1alpha1
+kind: alertingrule.monitoring.openshift.io
+metadata:
+  name: custom-alerts
+  namespace: openshift-monitoring
+spec:
+  groups:
+  - name: CustomGroup
+    rules:
+    - alert: CustomAlert
+      expr: "vector(1)"
+      for: 15m
+      labels:
+        custom: "true"
+```
+
+A new controller in the cluster-monitoring-operator will generate
+`PrometheusRule` resources based on the `AlertingRule` instances in the
+`openshift-monitoring` namespace.  These will be consumed by the platform
+prometheus-operator instance like any other `PrometheusRule`.  All user-defined
+alerts will be labeled identifying them as such -- likely using the existing
+`openshift_io_alert_source` label.
```

Comment:
what would be the value?

Reply:
Good question. The existing label is `openshift_io_alert_source=platform`, so maybe something like `openshift_io_alert_source=platform/user`? Where is that value normally viewed or used?

Reply:
Currently the label is only useful for alert routing (e.g. in the Alertmanager configuration) to identify "platform" vs. "user-defined" alerts.

Reply:
Updated this to be more concrete. The proof-of-concept branch uses: `openshift_io_alert_source=platform/user`